### PR TITLE
exp/services/ledgerexporter: Add guide for contributors/developers

### PR DESCRIPTION
**Pull request by urvisavla**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Add a guide for developers and contributors to complement the [Installation and Usage](https://github.com/stellar/go/blob/master/exp/services/ledgerexporter/README.md) guide.

### Why
Fixes [HUBBLE-457](https://stellarorg.atlassian.net/browse/HUBBLE-457)

### Known limitations

[HUBBLE-457]: https://stellarorg.atlassian.net/browse/HUBBLE-457?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+
+# Ledger Exporter Developer Guide
+The ledger exporter is a tool to export Stellar network transaction data to cloud storage in a way that is easy to access.
+
+## Prerequisites
+This document assumes that you have installed and can run the ledger exporter, and that you have familiarity with its CLI and configuration. If not, please refer to the [Link to Admin Guide]
+
+## Goal
+The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to cloud storage.
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:9]:
```suggestion
The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to a configurable remote data store, such as cloud blob storage.
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ - Minimize network usage to export
+ - Make it easy and fast to search for a specific ledger or ledger range
+
+## Architecture
+To achieve its goals, the ledger exporter uses the following architecture, which consists of the 3 main components:
+- Captive-core [Link to existing documentation on captive-core basics] to extract raw transaction metadata from the Stellar Network
+- Export manager to bundles and organizes the ledgers to get them ready for export
+- The cloud storage plugin writes to the cloud storage. This is specific to the type of cloud storage, GCS in this case.
+
+[Insert Architecture diagram here]
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:20]: here's a diagram from the eng talk that may be reusable, still lines up with current day

galexie

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to cloud storage.
+ - Use cloud storage optimally
+ - Minimize network usage to export
+ - Make it easy and fast to search for a specific ledger or ledger range
+
+## Architecture
+To achieve its goals, the ledger exporter uses the following architecture, which consists of the 3 main components:
+- Captive-core [Link to existing documentation on captive-core basics] to extract raw transaction metadata from the Stellar Network
+- Export manager to bundles and organizes the ledgers to get them ready for export
+- The cloud storage plugin writes to the cloud storage. This is specific to the type of cloud storage, GCS in this case.
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:18]: Maybe mention the `DataStore` interface/package as the aspect that provides pluggability? and `GCSDataStore` as one implementation.

**Review by sreuland — APPROVED** [review]
lgtm!

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+## Data Format
+- Ledger exporter uses a compact and efficient data format called XDR (External Data Representation), which is a compact binary format. The captive-core emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundle multiple LedgerCloseMeta's into a single object using a custom XDR struct called LedgerCloseMetaBatch which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).
+
+- The metadata for the same batch is also stored with for each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 
+
+- Objects are compressed before uploading using the zstd (zstandard) compression algorithm to optimize network usage and storage needs.
+
+## Data Storage (GCS)
+- The source for the cloud storage plugin is in the [support](https://github.com/stellar/go/tree/master/support/datastore) package. Any new storage plugin must implement the interface defined in [datastore.go](https://github.com/stellar/go/blob/master/support/datastore/datastore.go).
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:30]:
```suggestion
- An example implementation of `DataStore` for GCS, Google Cloud Storage. This plugin is located in the [support](https://github.com/stellar/go/tree/master/support/datastore) package. 
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+- Build: To build the Docker container, use the provided [Makefile](https://github.com/stellar/go/exp/services/ledgerexporter/Makefile). Simply run make `make docker-build` to build a new container after making any changes.
+
+- Run: For instructions on running the Docker container, refer to the Admin Guide [Link to Admin Guide].
+
+- Test: To test the Docker container, refer to the [docker-test](https://github.com/stellar/go/blob/master/exp/services/ledgerexporter/Makefile) command for an example of how to use the [GCS emulator](https://github.com/fsouza/fake-gcs-server) for local testing. 
+
+## Adding support for a new storage type
+To add support for a new storage type (e.g. AWS S3), follow these steps:
+
+- Implement the interface defined in [datastore.go](https://github.com/stellar/go/blob/master/support/datastore/datastore.go).
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:44]:
```suggestion
- A data storage plugin must implement the [DataStore](https://github.com/stellar/go/blob/master/support/datastore/datastore.go) interface.
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+## Build, Run and Test using Docker
+The Dockerfile contains all the necessary dependencies (e.g., Stellar-core) required to run the ledger exporter. 
+- Build: To build the Docker container, use the provided [Makefile](https://github.com/stellar/go/exp/services/ledgerexporter/Makefile). Simply run make `make docker-build` to build a new container after making any changes.
+
+- Run: For instructions on running the Docker container, refer to the Admin Guide [Link to Admin Guide].
+
+- Test: To test the Docker container, refer to the [docker-test](https://github.com/stellar/go/blob/master/exp/services/ledgerexporter/Makefile) command for an example of how to use the [GCS emulator](https://github.com/fsouza/fake-gcs-server) for local testing. 
+
+## Adding support for a new storage type
+To add support for a new storage type (e.g. AWS S3), follow these steps:
```
Comment (sreuland) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:42]:
```suggestion
Support for different data storage types are encapsulated as 'plugins', which are implementation of `DataStore` interface in a go package. To add a data storage plugin based on a new storage type (e.g. AWS S3), follow these steps:
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ - Make it easy and fast to search for a specific ledger or ledger range
+
+## Architecture
+To achieve its goals, the ledger exporter uses the following architecture, which consists of the 3 main components:
+- Captive-core to extract raw transaction metadata from the Stellar Network.
+- Export manager to bundles and organizes the ledgers to get them ready for export.
+- The cloud storage plugin writes to the cloud storage. This is specific to the type of cloud storage, GCS in this case.
+
+
+![ledgerexporter-architecture](https://github.com/urvisavla/go/assets/30014929/8fcc4733-6e35-4de7-ad06-0a27515364cd)
```
Comment (chowbao) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:21]: Should this asset be added to the go monorepo instead of your personal one?

**Review by chowbao — APPROVED** [review]

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+
+## Data Format
+- Ledger exporter uses a compact and efficient data format called XDR (External Data Representation), which is a compact binary format. The captive-core emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundle multiple LedgerCloseMeta's into a single object using a custom XDR struct called LedgerCloseMetaBatch which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).
```
Comment (Shaptic) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:25]:
```suggestion
- Ledger exporter uses a compact and efficient data format called [XDR](https://developers.stellar.org/docs/learn/encyclopedia/data-format/xdr) (External Data Representation), which is a compact binary format. A Stellar Captive Core instance emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundles multiple `LedgerCloseMeta`'s into a single object using a custom XDR structure called `LedgerCloseMetaBatch` which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+- The metadata for the same batch is also stored with for each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go).
```
Comment (Shaptic) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:27]:
```suggestion
- The metadata for the same batch is also stored alongside each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+![ledgerexporter-architecture](https://github.com/urvisavla/go/assets/30014929/8fcc4733-6e35-4de7-ad06-0a27515364cd)
+
+
+## Data Format
+- Ledger exporter uses a compact and efficient data format called XDR (External Data Representation), which is a compact binary format. The captive-core emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundle multiple LedgerCloseMeta's into a single object using a custom XDR struct called LedgerCloseMetaBatch which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).
+
+- The metadata for the same batch is also stored with for each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 
+
+- Objects are compressed before uploading using the zstd (zstandard) compression algorithm to optimize network usage and storage needs.
```
Comment (Shaptic) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:29]:
```suggestion
- Objects are compressed before uploading using the [zstd](http://facebook.github.io/zstd/) (zstandard) compression algorithm to optimize network usage and storage needs.
```

**Review by Shaptic — APPROVED** [review]
Nice! :clap: I just added some minor notes about typos and cross-references :+1:

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md (hunk 7 above)
Comment (urvisavla) [reply exp/services/ledgerexporter/DEVELOPER_GUIDE.md]: done

**Review by tamirms — APPROVED** [review]

**Pull request merged**